### ISSUE TITLE
IgxGridBaseDirective not correctly exported giving angular compile time error

### ISSUE THREAD

**Issue by fexxdev** [IgniteUI/igniteui-angular#13265]
## Description  
Hi everyone

We're using IGX in an enterprise app and i need some help.
I've been working on migrating to angular 16, but ever since i had trouble with this specific import and i don't know how to fix it.

Webstom automatically imports `IgxGridBaseDirective `from `@infragistics/igniteui-angular/lib/grids/grid-base.directive`
And if i press F12 it's correctly there

Until i compile

Then i have two errors (because i use it in two different directives), that says: 

`Module not found: Error: Package path ./lib/grids/grid-base.directive is not exported from package C:\blabla\node_modules\@infragistics\igniteui-angular`

Also your [documentation](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/igxgridcomponent.html#clipboardOptions) still redirects to [this](https://github.com/IgniteUI/igniteui-angular/blob/53312e306a/projects/igniteui-angular/src/lib/grids/grid-base.directive.ts#L318) which is exactly what i need but doesn't work on ng16.

 * igniteui-angular version: igniteui-angular-core: 16.1.0; igniteui-angular: 16.0.12
 * browser: Not needed

## Steps to reproduce  

1. Import IgxGridBaseDirective in a component

## Result  
What is the actual result after following the steps to reproduce?

Error while compiling (build watch and prod aswell)

## Expected result  
What is the expected result after following the steps to reproduce?

No compiler errors while compiling.

## Attachments  
Attach a sample if available, and screenshots, if applicable.  
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/139c7142-9b2e-4dc3-a7af-6390863aba87)
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/2376dc3f-a624-44b1-84f8-a14c5800619c)
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/86f5ce07-411c-4c8b-badb-07b0de90221e)

**Comment by kdinev** [comment]
@fexxdev The `IgxGridBaseDirective` was not intended for public use and was previously exported by accident. With `16.0.0` we've removed the export. If you are using a single type of grid, e.g. `IgxGridComponent` then you can use the `IgxGridComponent` type directly. If your grid can be of multiple types, e.g. `IgxGridComponent` and `IgxTreeGridComponent`, then the public common interface to use is `GridType`.

**Comment by fexxdev** [comment]
Hi @kdinev thanks for the reply!
I need some help as im upgrading to igx 16 and this was some legacy code that we used in our application.
This class that im working on is basically a wrapper for the grid state, to unify behaviors between your grid and what we need.
It implements IgxGridStateDirective, can you please confirm that this is intended for public use?

Then to satisfy that dependency i have in the ctor 
`@Host() @Optional() @Inject(IGX_GRID_BASE) private gridBaseDirective: IgxGridBaseDirective,`
 But you said that it is not intended for public use
Altough the super needs it as i can't access grid using this in the super ctor
`IgxGridStateDirective.constructor(grid: GridType, viewRef: ViewContainerRef, envInjector: EnvironmentInjector, injector: Injector)`

Lastly, the only thing that prevents me from upgrading is this piece of code
`this.gridBaseDirective.columnInit.pipe(...`
...
`column.filterable = customColumns.filterable;`
 `column.sortable = customColumns.sortable;`
`column.formatter = customColumns.formatter;`
`column.filters = customColumns.filters;`
`column.sortStrategy = customColumns.sortStrategy;`
`column.cellClasses = customColumns.cellClasses;`
`column.cellStyles = customColumns.cellStyles;`
...

we were able to listen to columnInit observable and react to it restoring the state of the grid. ColumnInit has since been removed and i dont know what else i can do to fix it. Any ideas?

Thanks for your patience
F.

**Comment by kdinev** [comment]
@fexxdev You can use `IgxGridComponent` instead of `IgxGridBaseDirective` type and both `autoGenerate` and `clipboardOptions` exist on it. I see that they are currently missing in the common public interface `GridType` and we will add them.

**Comment by fexxdev** [comment]
Hey @kdinev thanks for your quick answer. I'll try again later this day with your tips. 
Anyway, what about columnInit?

**Comment by kdinev** [comment]
@fexxdev `columnInit` also exists on the `IgxGridComponent`. The `IgxGridComponent` extends `IgxGridBaseDirective` and has everything that is a public member of it.

**Comment by fexxdev** [comment]
@kdinev what about options? 
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/6a325783-f786-4b53-be62-3cfcd3dd8e2e)

I can't find it in IgxGridComponent..

**Comment by rkaraivanov** [comment]
@fexxdev 
That's a property of the state directive and has nothing to do with the grid

**Comment by damyanpetev** [comment]
> @damyanpetev I'm including you to this conversation to comment on whether there's any reason for these two to not be declared in the `GridType`. `columnInit` is also a missing output in the interface.

It's both missing props and exposing some not intended. There's a task logged for this already https://github.com/IgniteUI/igniteui-angular/issues/12342, though it's a bit icy atm :)

> > @fexxdev That's a property of the state directive and has nothing to do with the grid
> 
> Yep, the problem is that if i extend IgxGridStateDirective i need to implement the super ctor, that needs GridType that was exposed in IgxGridBaseDirective but that is what i had to remove because its not exposed anymore ![image](https://user-images.githubusercontent.com/26311786/254916443-64902b11-d42e-4837-9ae6-5feee1866bbe.png)

While we're locked in using using tokens and the type interface around the library code in order to avoid circular references between components, your consuming apps shouldn't be. 
The suggestion to use [`IgxGridComponent`](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/IgxGridComponent.html) is sound - as visible in the API it implements the [`GridType`](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/interfaces/GridType.html) be that through inheritance from base or explicitly. So the Grid component should be directly assignable anywhere the interface is required (including the `ctor`) and also type asserted from interface properties (knowing the usage). In fact applies to all grids, which is that interface's point.

For the benefit of others, if you're extending the state directive, keep in mind you can inherit the `ctor` without overriding so you can save yourself the trouble of setting up the injection again. Either way should work though.

Also if your issue is with accessing the `grid` property, again that should easily type assert `as IgxGridComponent` and even easier when setting up through the constructor.
I feel like showing is better than telling, so:
```ts
@Directive({ selector: '[customGridState]' })
export class CustomGridStateDirective extends IgxGridStateDirective implements OnInit {

    public override grid: IgxGridComponent;

    ngOnInit() {
      this.grid.autoGenerate = false;
      this.grid.columnInit.pipe(takeUntil(this.destroy$)).subscribe(col => {
        col.filterable = true;
      });    }
}
```
Or looking at the extra dependencies injected:
```ts
@Directive({ selector: '[customGridState]' })
export class CustomGridStateDirective extends IgxGridStateDirective {
  constructor(
    @Host() @Inject(IgxGridComponent) public override grid: IgxGridComponent,
    viewRef: ViewContainerRef, envInjector: EnvironmentInjector, injector: Injector
  ) {
    super(grid, viewRef, envInjector, injector);
    this.grid.autoGenerate = false;
    this.grid.columnInit.pipe(takeUntil(this.destroy$)).subscribe(col => {
      col.filterable = true;
      // ...
    });
  }
}
```
Hope this is useful and here's a StackBlitz with the above in use: https://stackblitz.com/edit/wtu5zx?file=src%2Fapp%2Fdirectives%2Fcustom-state.directive.ts,src%2Fapp%2Fgrid%2Fgrid-save-state%2Fgrid-state.component.html

**Comment by fexxdev** [comment]
@damyanpetev yep that cleared up many doubts i had :D Thanks you so much for the example

**Comment by fexxdev** [comment]
@damyanpetev 
i've noticed another thing
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/46e510c4-b4a9-45d9-8bc1-35ab9b2d604d)
this is from your angular docs
But in reality
![image](https://github.com/IgniteUI/igniteui-angular/assets/26311786/5601c43d-3519-4c54-8b2d-a0f2a01ec095)

**Comment by fexxdev** [comment]
Hi guys, any update on the last report? thanks!

**Comment by fexxdev** [comment]
@damyanpetev i'll update this issue asap, im currently testing out the whole app after the ng16 and igx16 upgrade to ensure everything works as expected. Thanks for your kind help and support though, much appreciated! 👍🏻
